### PR TITLE
[#60] Throwing problems with http headers

### LINKED ISSUE #60: Throwing problems with http headers
**Is your feature request related to RFC7807? Please describe.**
Zalando library does not provide an easy way to define custom headers when throwing `ThrowableProblem` - problem itself describes only response body. This becomes a challenge, when developer wants to have both - `application/problem+json` and custom headers (e.g for http 429 + `Retry After` header).

One possible walk around is to throw `WebApplicationException` with `Response` where we can set headers and provide `Problem` as entity/body:
```
throw new WebApplicationException(
    Response
        .status(Response.Status.BAD_REQUEST)
        .header("X-Special-Header", "ABC")
        .type("application/problem+json")
        .entity(
            Problem.builder()
                .withStatus(Status.BAD_REQUEST)
                .withTitle("Problem title")
                .with("custom_field", "CustomValue")
                .build()
            )
        .build()
);
```

This works pretty well, but such `WebApplicationException` will skip all ExceptionMappers and our PostProcessors, so it will not be enhanced (`instance` default value, mdc fields injection) nor logged, so does not provide consistent developer experience.

**Describe the solution you'd like**
Provide new abstraction `HttpProblem` + nice builder + ExceptionMapper that will handle this type properly. Possible usage:
<pre>
throw HttpProblem.builder()
    .withStatus(Status.BAD_REQUEST)
    .withTitle("Problem title")
    <b>.withHeader("X-Special-Header", "ABC")</b>
    .with("custom_field", "CustomValue")
    .build();
</pre>

Zalando's `ProblemBuilder` class is final, so we have to create our own copy with headers functionality.

This should produce exactly the same response than WebApplicationException in the example above, but transformed by PostProcessors (mdc fields, logging etc):
```
HTTP/1.1 400 Bad Request
X-Special-Header: ABC
Content-Type: application/problem+json
        
{
  "title": "Problem title",
  "status": 400,
  "custom_field": "CustomValue",
  "mdc_field": "123"
}
```

This could also be a step towards getting rid of Zalando lib dependency.

### PR DESCRIPTION
**Pull request by lwitkowski**
See #60 for details and rationale. HttpProblem will probably slightly change in the next PRs (maybe we will decide to extract interface as Zalando did?).

Closes #60

### CODE REVIEW

File: runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Test
+    void copyBuilderShouldCopyAllFields() {
+        HttpProblem original = sampleHttpProblem();
+
+        HttpProblem copy = HttpProblem.builder(original).build();
+
+        assertThat(copy)
+                .isEqualToComparingFieldByField(original);
```
Comment (pazkooda) [inline runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java:41]: `.isEqualToComparingFieldByField` is deprecated?

File: runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package com.tietoevry.quarkus.resteasy.problem;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.net.URI;
+import org.junit.jupiter.api.Test;
+import org.zalando.problem.Status;
+
+class HttpProblemTest {
```
Comment (pazkooda) [inline runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java:9]: Possibly add test on reserved properties throwing `IllegalArgExc`.

**Review by pazkooda — APPROVED** [review]
LGTM (with comments 🙂 )

File: runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java (hunk 1 above)
Comment (lwitkowski) [reply runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java:41]: Nope, it's not, at least not in assertj version we're using.

File: runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java (hunk 2 above)
Comment (lwitkowski) [reply runtime/src/test/java/com/tietoevry/quarkus/resteasy/problem/HttpProblemTest.java:9]: Good point, adding now.

**Pull request merged**